platform/mbed_assert.h

```diff
@@ -49,6 +49,65 @@ do {                                                     \
 } while (0)
 #endif
 
+
+#define _MBED_CONCAT_(a, b) a##b
+#define _MBED_CONCAT(a, b) _MBED_CONCAT_(a, b)
+#define _MBED_ASSERTION _MBED_CONCAT(_MBED_ASSERTION_AT_, __LINE__)
+
+/** MBED_STATIC_ASSERT
+ *  Declare compile-time assertions, results in compile-time error if condition is false
+ *
+ *  The assertion acts as a declaration that can be placed at file scope, in a
+ *  code block (except after a label), or as a member of a C++ class/struct/union.
+ *
+ *  @code
+ *  MBED_STATIC_ASSERT(MBED_LIBRARY_VERSION >= 120,
+ *          "The mbed library must be at least version 120");
+ *
+ *  int main() {
+ *      MBED_STATIC_ASSERT(sizeof(int) >= sizeof(char),
+ *              "An int must be larger than a char");
+ *  }
+ *  @endcode
+ */
+#ifdef __cplusplus
+#if __cplusplus >= 201103L || __cpp_static_assert >= 200410
+#define MBED_STATIC_ASSERT(expr, msg) static_assert(expr, msg)
+#elif defined(__GXX_EXPERIMENTAL_CXX0X__) && __GNUC__ >= 4 && __GNUC_MINOR__ >= 3
+#define MBED_STATIC_ASSERT(expr, msg) __extension__ static_assert(expr, msg)
+#else
+#define MBED_STATIC_ASSERT(expr, msg) enum {_MBED_ASSERTION = sizeof(char[(expr) ? 1 : -1])}
+#endif
+#else
+#if __STDC_VERSION__ >= 201112L
+#define MBED_STATIC_ASSERT(expr, msg) _Static_assert(expr, msg)
+#elif defined(__GNUC__) && __GNUC__ >= 4 && __GNUC_MINOR__ >= 6 && !defined(__CC_ARM)
+#define MBED_STATIC_ASSERT(expr, msg) __extension__ _Static_assert(expr, msg)
+#else
+#define MBED_STATIC_ASSERT(expr, msg) enum {_MBED_ASSERTION = sizeof(char[(expr) ? 1 : -1])}
+#endif
+#endif
+
+/** MBED_STRUCT_STATIC_ASSERT
+ *  Declare compile-time assertions, results in compile-time error if condition is false
+ *
+ *  Unlike MBED_STATIC_ASSERT, MBED_STRUCT_STATIC_ASSERT can and must be used
+ *  as a member of a C/C++ class/struct/union.
+ *
+ *  @code
+ *  struct thing {
+ *      MBED_STATIC_ASSERT(2 + 2 == 4,
+ *              "Hopefully the universe is mathematically consistent");
+ *  };
+ *  @endcode
+ */
+#if defined(__cplusplus)
```

Comment:
This is incorrect, a C++ compiler can treat a structure as a C one if the structure is imported  as `extern "C"`.